dali/pipeline/operator/arg_helper.h

```diff
@@ -45,32 +52,49 @@ class ArgValue {
     return *this;
   }
 
-  inline bool IsTensor() const { return tensor_vector_ != nullptr; }
+  inline bool IsInput() const { return tensor_vector_ != nullptr; }
+  inline bool IsSet() const { return is_set_; }
+
+  template <int ndim = DynamicDimensions>
+  inline const TensorView<StorageCPU, T, ndim> Tensor(int sample_index) {
+    ASSERT(IsInput());
+    return view<T, ndim>(tensor_vector_[sample_index]);
+  }
+
+  inline const T &Value() {
+    return value_;
+  }
 
-  inline const T &operator[](Index index) {
-    if (IsTensor()) {
+  inline const T &operator[](int sample_index) {
+    if (IsInput()) {
 #if DALI_DEBUG
-      DALI_ENFORCE(index < static_cast<Index>(tensor_vector_->ntensor()));
+      DALI_ENFORCE(sample_index < static_cast<int>(tensor_vector_->ntensor()));
 #endif
-      return (*tensor_vector_)[index].data<T>()[0];
+      return (*tensor_vector_)[sample_index].data<T>()[0];
     } else {
       return  value_;
     }
   }
 
-  inline const TensorList<GPUBackend> *AsGPU(cudaStream_t stream) {
-    DALI_ENFORCE(IsTensor());
-    if (!gpu_) {
+  inline const TensorList<GPUBackend> &AsGPU(cudaStream_t stream) {
+    DALI_ENFORCE(IsInput());
+    if (!gpu_)
       gpu_.reset(new TensorList<GPUBackend>());
```

Comment:
Why we need it to be a unique_ptr? If you don't set any size to TensorList<GPUBackend> it won't allocate, doesn't it?

Reply:
1. TensorList imposes restrictions on assignment and construction - a thing which I wanted to avoid here.
2. Even when empty, it's a heavy object - this wrapper was intended for use with any arguments, including scalars - in this case carrying the weight of an object which is never used is unjustifiable - especially given the fact that allocating a GPU tensor list is so heavy as to totally eclipse the cost of dynamically allocating the TensorList itself on regular heap.